Kerbos services and more sasl authorization mechanism support

axeisghost opened the pull request:
The idea is from [Impala](https://github.com/cloudera/impyla/blob/255b07ed973d47a3395214ed92d35ec0615ebf62/impala/_thrift_api.py#L152-L160). I added more check for the mechanism and also allow user to put in the kerbos services name if they want to use the service.

jingw commented on pyhive/hive.py:
```diff
@@ -66,11 +66,14 @@ def connect(*args, **kwargs):
     """Wraps a Thrift session"""
 
-    def __init__(self, host, port=10000, username=None, database='default', auth='NONE',
+    def __init__(self, host, kerberos_service_name=None, port=10000, username=None, password=None, database='default', auth='NONE',
```
Inserting arguments is backwards incompatible for anyone not using named arguments)

jingw commented on pyhive/hive.py:
```diff
@@ -79,20 +82,29 @@ def __init__(self, host, port=10000, username=None, database='default', auth='NO
             # NOSASL corresponds to hive.server2.authentication=NOSASL in hive-site.xml
             self._transport = thrift.transport.TTransport.TBufferedTransport(socket)
-        elif auth == 'NONE':
+        elif auth.upper() in ['LDAP', 'PLAIN', 'GSSAPI', 'NONE']:
```
PLAIN and GSSAPI aren't valid values for hive.server2.authentication (see docstring). Can we just have one supported way? (no auto-remapping and no auto-casing)

jingw commented on pyhive/hive.py:
```diff
@@ -79,20 +82,29 @@ def __init__(self, host, port=10000, username=None, database='default', auth='NO
+                sasl_client.setAttr('host', host)
+                sasl_client.setAttr('service', kerberos_service_name)
+                if auth.upper() in ['PLAIN', 'LDAP']:
+                    sasl_client.setAttr('username', user)
```
I don't see `user` defined here. Also what happens if `user` is passed as a unicode string?

jingw commented:
- Could you fill out Dropbox's CLA at https://opensource.dropbox.com/cla/?
- How was this tested? If I recall correctly `sasl` was picking about taking byte strings instead of unicode strings, so the code here looks like it'll break (we have __future__ unicode_literals).
- Could you get this auth mode running on Travis CI? The build matrix is in .travis.yml and the setup script is https://github.com/dropbox/PyHive/blob/master/scripts/travis-install.sh

axeisghost commented:
Sorry for the late reply. I just signed the CLA agreement and I also updated the branch.

I modified my code according to the suggestion you mentioned. I checked my code again and also set up the Travis CI building test. I apologized for my careless mistakes.
For the hive.server2.authentication mode you mentioned, I tried to map the mode of [hive.server2.authentication](https://cwiki.apache.org/confluence/display/Hive/Setting+Up+HiveServer2#SettingUpHiveServer2-Authentication/SecurityConfiguration) to the mode of [sasl library](https://github.com/cloudera/python-sasl) from cloudera. The sasl library does not have the documentation of the authentication mode it supported but another project they have, Impala provides some hints for it and I just follow the [usage](https://github.com/cloudera/impyla/blob/255b07ed973d47a3395214ed92d35ec0615ebf62/impala/_thrift_api.py#L120-L166) in there.

The test result of the branch that I fixed is [here](https://travis-ci.org/axeisghost/PyHive/builds/145960160).

jingw commented:
Thanks. The reason I'm asking for test coverage is that I need to know this code works and is supportable, but I don't have a LDAP/Kerberos setup myself.

Two possible ways to test this:
- Integrate into build matrix in .travis.yml. Have some kind of HIVE_AUTH_MODE variable that controls what happens in the setup scripts, including what makes it to hive-site.xml.
- Write a targeted unit test that edits hive-site.xml and restarts hiveserver2 (and afterwards resets hive-site.xml)

axeisghost commented:
Thanks for the suggestions. That is what we are looking at right now. We will push as soon as we get progress.

axeisghost commented:
We read through the test and environment setup code and have a general idea about the test process. However, we do not know which command line will start the hiveServer2. After copy paste the hive-site.xml, the script will install hive-server2 but it goes to run the setup script right after installation. Can we assume that the apt-get install step will start up the hive-server2? When we restart the hiveserver2, can we just run `hive --service hiveserver2`? Thank you.

jingw commented:
The apt-get command does start HiveServer2.

To restart hiveserver2, you can do `sudo service hive-server2 start`
https://www.cloudera.com/documentation/cdh/5-0-x/CDH5-Installation-Guide/cdh5ig_hiveserver2_start_stop.html

axeisghost commented:
LDAP mech is actually included in the PLAIN mech in libsasl2-dev, documentation [here](https://cyrusimap.org/mediawiki/index.php/Cyrus_SASL#Plugins_.28SASL_Mechanisms.29). LDAP test case was added but actually there is no extra branch specified for LDAP.
KERBEROS mech branch is the only extra branch in this Pull Request.

axeisghost commented:
And it seems the '--flake8' options in setup.cfg for py.test was not recognized. I believe this is a problem from travis CI, I tried to build master branch and it failed for the same reason.

jingw commented:
I'll get back to you tomorrow on the flake8 stuff. I'm guessing I should just pin all my package versions.

jingw commented:
Looks like it gets past the pytest invocation now, but the build hangs
https://travis-ci.org/dropbox/PyHive/jobs/147552364

axeisghost commented:
[](https://travis-ci.org/dropbox/PyHive/jobs/144597529)
It seems the master branch has the same problem too. Newer version of flake8 will stall some of the check. I am trying to ignore more files and see how will it work.

jingw commented:
I couldn't figure out why the latest version of pytest-flake8 / flake8 breaks the build (actually couldn't repro on my dev box, only happens on Travis). I've pinned the packages to get master passing again.

If you rebase it should actually test now. Thanks, and sorry for the delay.

jingw commented on pyhive/tests/test_hive.py:
```diff
@@ -139,3 +139,30 @@ def test_no_result_set(self, cursor):
+    __test__ = True
+
+    def test_ldap_connection(self):
+        import subprocess
```
imports should be at the top of the file

jingw commented on pyhive/tests/test_hive.py:
```diff
@@ -139,3 +139,30 @@ def test_no_result_set(self, cursor):
+
+@unittest.skipIf(sys.version_info.major == 3, 'Hive not yet supported on Python 3')
+class TestHiveAuth(unittest.TestCase):
+    __test__ = True
```
I believe this is unnecessary

jingw commented on pyhive/tests/test_hive.py:
```diff
@@ -139,3 +139,30 @@ def test_no_result_set(self, cursor):
+        import subprocess
+        import os
+        rootdir = os.environ['TRAVIS_BUILD_DIR']
+        subprocess.call(['sudo', 'cp', rootdir + '/scripts/travis-conf/hive/hive-site-ldap.xml',
```
- Prefer subprocess.check_call, which throws an exception if the command fails
- Prefer os.path.join(rootdir, 'scripts', 'travis-conf', ...)

jingw commented on pyhive/tests/test_hive.py:
```diff
@@ -139,3 +139,30 @@ def test_no_result_set(self, cursor):
+        subprocess.call(['sudo', 'cp', rootdir + '/scripts/travis-conf/hive/hive-site-ldap.xml',
+                         '/etc/hive/conf/hive-site.xml'])
+        subprocess.call(['sudo', 'service', 'hive-server2', 'restart'])
+        subprocess.call(['sleep','10'])
```
prefer doing this without a subprocess: time.sleep(10)

jingw commented on pyhive/tests/test_hive.py:
```diff
@@ -139,3 +139,30 @@ def test_no_result_set(self, cursor):
+        connection = hive.connect(host=_HOST, username='existing', auth='LDAP',
+                                  configuration={'mapred.job.tracker': 'local'},
+                                  password='testpw')
+        cursor = connection.cursor()
```
Please close the connection/cursor to avoid leaks: https://github.com/dropbox/PyHive/blob/master/pyhive/tests/dbapi_test_case.py#L22

jingw commented on pyhive/tests/test_hive.py:
```diff
@@ -139,3 +139,30 @@ def test_no_result_set(self, cursor):
+        self.assertEqual(cursor.fetchone(), (1,))
+        self.assertEqual(cursor.rownumber, 1)
+        self.assertIsNone(cursor.fetchone())
+        subprocess.call(['sudo', 'cp', rootdir + '/scripts/travis-conf/hive/hive-site.xml',
```
This cleanup should technically be in a try/finally block

jingw commented on pyhive/tests/test_hive.py:
```diff
@@ -25,6 +28,9 @@ class TestHive(unittest.TestCase, DBAPITestCase):
     def connect(self):
         return hive.connect(host=_HOST, configuration={'mapred.job.tracker': 'local'})
 
+    def connectWithKeyWordArgs(self, *args, **kwargs):
```
This wrapper seems unecessary

jingw commented on pyhive/hive.py:
```diff
@@ -67,32 +67,49 @@ class Connection(object):
+            if auth.upper() == 'KERBEROS' or kerberos_service_name:
+                # KERBEROS mode in hive.server2.authentication is
+                # GSSAPI in sasl library
+                auth = 'GSSAPI'
```
This line isn't covered. I'll probably just remove when merging.

axeisghost replied:
I updated other two lines. This line is for Kerberos service that we need. It is a little bit hard to set up hive on kerberos. We are still working on it. I will let you know with any progress we made. Thanks.

jingw commented on pyhive/hive.py:
```diff
@@ -67,32 +67,49 @@ class Connection(object):
+                auth = 'GSSAPI'
+            if password is None:
+                if auth == 'LDAP':
+                    password = ''
```
This line isn't covered. Can we just make password mandatory for LDAP?

jingw commented:
Awesome, tests passing :+1:

Dawny33 commented:
Any update on this?  Would be great to have Kerberos auth support :)

alexwoolford commented:
Hi @jingw. I'd like to use the Kerberos functionality provided by @axeisghost's PR. I would imagine there are an increasing number of folks (e.g. [@wanggaohang](https://github.com/dropbox/PyHive/issues/47)) who'd like to use PyHive to query Kerberized clusters. Do you plan to merge and release this?

mistercrunch commented:
+1

360d-io-labs commented:
+1 Kerberos is becoming a basic standard

x-x-l commented:
+2354123

PaulLiang1 commented:
+1

evinhas commented:
Hi,

It would be great that pyhive will include this feature: kerberos is used in most of the cloudera clusters, so most of the people must use kerberos for getting connected to Hive.

jingw commented:
I gave up trying to find time to personally set up and test Kerberos, so I'm going ahead with the tried and tested merge & pray strategy.

alexwoolford commented:
Thanks for merging this, @jingw . FWIW, I used @axeisghost 's PR to connect to Kerberized Hive and it worked well for me.
